### ISSUE TITLE
`<algorithm>`: Should `_Rng_from_urng` use `make-unsigned-like-t`?

### ISSUE THREAD

**Issue by hewillk** [microsoft/STL#2884]
```cpp
#include <vector>
#include <random>
#include <ranges>
#include <algorithm>

int main() {
  std::vector<std::size_t> v(10);
  std::ranges::sample(
    std::views::iota(0ULL, 42ULL),
    v.begin(), 3, std::mt19937{std::random_device{}()});
}
```
https://godbolt.org/z/bMredxf58

In the above example, the `difference_type` of `iota_view` is `_Signed128`, which makes the instantiation of `_Rng_from_urng` fail since it uses `make_unsigned_t` which cannot work with integer-class types.

https://github.com/microsoft/STL/blob/ef62d3fa0b8e4e2406b9bb74e916e1ca8a1df802/stl/inc/xutility#L5696-L5700

It's worth noting that gcc will only work with `-std=gnu++20` because `__int128` is an integer type _only_ in gnu-mode. [gcc will also fail the `static_assert` under `-std=c++20`](https://godbolt.org/z/Gn6qaMToW), but this assertion comes from the implementation of `std::sample`, so there is no reference value because `std::sample` is _not_ for C++20 iterators system.

In [[alg.random.sample]](https://eel.is/c++draft/alg.random.sample), I don't see any requirement that the above code violates, so should I consider it a bug?

**Comment by CaseyCarter** [comment]
Yes, this is a bug. It looks like `size_t _Bmask` is going to be limiting, but we can't change the layout of `_Rng_from_urng` - we'll probably need a partial specialization `_Rng_from_urng<_Signed128, _Urng>`.

**Comment by hewillk** [comment]
Note that `ranges::shuffle` has the same problem since it also uses `_Rng_from_urng`.

```cpp
#include <algorithm>
#include <ranges>
#include <vector>
#include <random>

int main() {
  std::vector<int> v;
  auto r = std::views::iota(0ULL, v.size())
         | std::views::transform([&v](auto i) -> auto& { return v[i]; });
  std::ranges::shuffle(r, std::mt19937{std::random_device{}()});
}
```

https://godbolt.org/z/89Wa15GhT

**Comment by MattStephanson** [comment]
The algorithms only need to generate numbers in the range [0,`size(r)`). As long as that can be represented by an integral type, then generating number in the larger, signed `difference_type` is unnecessary.

>  It looks like size_t _Bmask is going to be limiting, but we can't change the layout of _Rng_from_urng

I think `_Bmask` is fine. It's a mask for the bits that can be generated in *one* call to the URBG. The allowed result types for a URBG are a subset of the integral types, so 64 bits should always be sufficient.

**Comment by StephanTLavavej** [comment]
Because `_Rng_from_urng` is an internal type that isn't persistently stored anywhere (it appears only as local variables), we can totally modify its definition and even its data member layout *as long as* we rename it to `_Rng_from_urng_2`. That will avoid any mix-and-match problems. We've done this in the past with internal parallel algorithms machinery.

This may be entirely moot - as @MattStephanson noted and @CaseyCarter agreed (in our weekly maintainer meeting), as `_Bmask` isn't limiting, there should be no need to change the layout and thus we can continue using the same name.

**Comment by frederick-vs-ja** [comment]
`_Rng_from_urng_v2` has been moved to `<random>` (#3157). Should we move it to `<algorithm>` in order to fix this issue?
